API documentation: Group properties by meaning

iwis opened the pull request:
## Description
Implements https://github.com/MudBlazor/MudBlazor/issues/2980. Currently I added only categories for MudComponentBase, MudTableBase, and MudTable.
I put `[Parameter]` attribute and `[Category(CategoryTypes.Table.Appearance)]` attribute in separate lines like in [Google Style Guide](https://google.github.io/styleguide/csharp-style.html#attributes). If the coding style is ok I will add categories for other components.

## How Has This Been Tested?
Tested visually.

<!-- All PR's should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

Screenshot:
![Table - grouping properties by category](https://user-images.githubusercontent.com/8080496/142764940-9fffb1b1-1124-40eb-a20a-368029932868.png)

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
✔️ The PR is submitted to the correct branch (`dev`).
✔️ My code follows the code style of this project.
✔️ I've added relevant tests.

iwis commented:
> I put `[Parameter]` attribute and `[Category(CategoryTypes.Table.Appearance)]` attribute in separate lines like in [Google Style Guide](https://google.github.io/styleguide/csharp-style.html#attributes). If the coding style is ok I will add categories for other components.

So instead of:
```
[Parameter] public int Elevation { set; get; } = 1;
```        
It will be:
```
[Parameter]
[Category(CategoryTypes.Table.Appearance)]
public int Elevation { set; get; } = 1;
```
@JonBunator, @Garderoben: Is this format ok? Sorry for bothering you, but if I knew this single thing, I could start editing other files :)

JonBunator commented:
> > I put `[Parameter]` attribute and `[Category(CategoryTypes.Table.Appearance)]` attribute in separate lines like in [Google Style Guide](https://google.github.io/styleguide/csharp-style.html#attributes). If the coding style is ok I will add categories for other components.
> 
> So instead of:
> 
> ```
> [Parameter] public int Elevation { set; get; } = 1;
> ```
> 
> It will be:
> 
> ```
> [Parameter]
> [Category(CategoryTypes.Table.Appearance)]
> public int Elevation { set; get; } = 1;
> ```
> 
> @JonBunator, @Garderoben: Is this format ok? Sorry for bothering you, but if I knew this single thing, I could start editing other files :)

Yes looks good. Thanks for doing this :)

henon commented:
How about non-component specific categories for the most common ones? Like `CategoryTypes.General.Appearance`

iwis commented:
> How about non-component specific categories for the most common ones? Like `CategoryTypes.General.Appearance`

I thought about it. The advantage is that when **adding a new component that uses only these categories**, you don't need to alter the `CategoryAttribute.cs` file.

The disadvantage is that when **adding a new property to an existing component Component1** you have to use both lists: `CategoryTypes.General` and `CategoryTypes.Component1` to choose a category.

It seems to me that adding a property to an existing component occurs more often than adding a new component? So in my opinion, it can better to have one list for each component to make adding properties easier. Additionally, this solution seems to me a little bit easier mentally.

But in fact I'm not sure which option is better.

If we do it as in your proposal then there is a question how many general categories should be. The following two categories:
```
CategoryTypes.General.Behavior
CategoryTypes.General.Appearance
```
or the following 4-5 categories:
```
CategoryTypes.General.Behavior
CategoryTypes.General.Appearance
CategoryTypes.General.Data
CategoryTypes.General.Validation
CategoryTypes.General.ClickAction
```
In the latter case, a developer adding a new component would have to edit the CategoryAttribute.cs file even less often, but on the other hand a developer adding a new property to an existing component would sometimes have categories in the CategoryTypes.General list that don't apply to that component and would therefore have a harder choice. So it is rather bad in my opinion.

I haven't contributed to any component, so in fact you should make the decision what is better. I was mainly looking from the user's point of view. If you will not know what to choose then probably it isn't so important, because you can always change it later automatically.

henon commented:
Yeah, I see the dilemma. I also thought that maybe a single big enum would probably be the easiest solution as it avoids exactly this dilemma: `[[Category(CategoryTypes.Behavior)]` 

What would speak against that? I would see no real problem with component specific categories to be in the same enum and when choosing from the enum you easily find what special categories other components might already use.

iwis commented:
I will reply in half an hour, there are some nuances that I want to describe.

iwis commented:
I proposed the single enum in issue https://github.com/MudBlazor/MudBlazor/issues/2980#issuecomment-970090130, however JonBunator preferred current solution over enum. I accepted the choice because again the current solutions seems to me mentally easier a little bit from the following reasons:

1. A developer adding a property to an existing component sees exactly what categories he can choose - there are no categories on the list that are not used or don't make sense for this component (with some exceptions), so selecting is easier. If there was one common enum, the developer might have the following dilemmas:

    a) Popup properties are grouped into these categories:
    - PopupBehavior in `MudMenu`
    - ListBehavior in `MudSelect` and `MudAutocomplete`
    - PickerBehavior in pickers

     I had trouble choosing a name for them. I thought that one category "PopupBehavior" is not good idea because pickers in [static mode](https://mudblazor.com/components/datepicker#static-mode) don't have popups. Maybe one common category ListBehavior is better? But is a calendar a list? So I made three categories with quite sensible meanings, I hope. Now, a developer adding the `MudDatePicker.AutoClose` property may have a problem with choosing a category: should I choose PickerBehavior or PopupBehavior category?

    b) Should the `MudChip.Command` property have ClickAction or Selecting category?
    c) Should the `MudBaseInput<T>.ReadOnly` property have Behavior or Editing category?
    d) Developer could sometimes choose mistakenly "Common" category, which is currently reserved for `MudComponentBase`.

2. Additionally, the definition of `CategoryTypes` is a good list of categories uses in components, and thus I think that you can manage them, add new categories, reorganize etc more easy. My suggestion with enum seems to me little chaotic. Someone might wonder what the "Editing" category means - I wanted to solve it with the `// used in MudTable` comment, but such comments can quickly become obsolete, because maybe this category will also be used in MudTreeView in the future?

So I explained to myself JonBunator's decision in this way. But again, I have no experience in editing components, so it's not up to me to decide. Maybe @JonBunator can add something because he voted for it not to be an enum.

henon commented:
Thanks for the comprehensive insight. I see you have thought this through and your argument makes a lot of sense. Let's do it just like you and Jonbunator agreed upon.

iwis commented:
I added categories for 93% properties of components whose documentation is available in the ["API" menu](https://dev.mudblazor.com/api/). There are about 50 properties left, but it will take me a few days, because I didn't think about them before - some of these properties are new. In addition, classes unavailable from the "API" menu, for example [MudClumn](https://dev.mudblazor.com/api/column), also don't have categories, but this is less important IMHO.

I changed nearly 100 files so the PR is large. I would like to avoid merge conflicts, because I don't have experience with team work on GitHub. Can the PR be added to dev now to avoid conflicts, and then I will add missing categories (at least the most important), or we can deal with it in another way?

iwis commented:
@JonBunator, @Garderoben, @henon:

After adding a test to this PR, I realized that instead of adding attributes in the source files of individual classes, we can define the following map in a separate file:

```C#
Dictionary<string, CategoryEnum> propertyCategories = new()
{
    ["MudTableBase.Bordered"] = CategoryEnum.Appearance,
    ["MudTableBase.Dense"] = CategoryEnum.Appearance,
    ["MudTable.ChildRowContent"] = CategoryEnum.Rows,
    ["MudBadge.ChildContent"] = CategoryEnum.Behavior,
    ...
};
```

or:

```C#
Dictionary<string, Dictionary<string, CategoryEnum>> propertyCategories = new()
{
    [nameof(MudTableBase)] = new()
    {
        [nameof(MudTableBase.Bordered)] = CategoryEnum.Appearance,
        [nameof(MudTableBase.Dense)] = CategoryEnum.Appearance,
        ...
    },

    [nameof(MudTable<object>)] = new()
    {
        [nameof(MudTable<object>.ChildRowContent)] = CategoryEnum.Rows,
        ...
    },

    [nameof(MudBadge)] = new()
    {
        [nameof(MudBadge.ChildContent)] = CategoryEnum.Behavior,
        ...
    },
    ...
};
```

Previously, I thought it was a worse solution, because I thought that if we define categories in a separate file, programmers will add them less often. But we can add a test which will require that for each `[Parameter]` a category in the above-mentioned map has to be added.

Do you prefer adding categories in a separate file, or we can leave it as it is i.e. in components source files?

henon commented:
I think adding the attributes in the components is easier to understand for everyone. But it is great to have a test case that will complain if you forget to add it.

iwis commented:
I added categories for all properties of all components whose documentation is available in the "Components" menu and "API" menu. A few properties belong to "Misc" category only because it is the best category for them in my opinion:
 - MudDialog.Options - because it can alter both behavior and appearance
 - MudDialogInstance.Options - as above
 - all 3 properties of MudElement

The MudRatingItem class and the MudPicker.AllowKeyboardInput property probably should be removed from the documentation, so I also haven't added categories for them.

henon commented:
Huge thanks, I know this was a lot of work!
